PyTAPS/framer.py

```diff
@@ -3,31 +3,137 @@
+                await self.start_waiter
+            else:
+                break
+        """
+        self.start_waiter = self.loop.create_future()
+        try:
+            await self.start_waiter
+        finally:
+            self.start_waiter = None
+
+    """ Function that blocks until framer is after send event
+    """
+    async def await_framer_send(self):
+        send_waiter = self.loop.create_future()
+        self.send_waiter_list.append(send_waiter)
+        try:
+            data = await send_waiter
+        finally:
+            send_waiter = None
+            return data
+
+    """ Function that blocks until framer is after receive event
+    """
+    async def await_framer_receive(self):
+        receive_waiter = self.loop.create_future()
+        self.receive_waiter_list.append(receive_waiter)
+        try:
+            context, data, eom = await receive_waiter
+        finally:
+            receive_waiter = None
+            return context, data, eom
+
+    """ Empty functions, to be implemented by the framer implementation
+    """
+    async def start(self, connection):
+        pass
+    async def new_sent_message(self, data, context, eom):
+        pass
+    async def handle_received_data(self, connection):
+        pass
+    async def stop(self):
+        pass
+
+    # Fire start event and the wait until framer replies
+    async def handle_start(self, connection):
+        self.loop.create_task(self.start(connection))
+        await self.await_framer_ready()
+        return
+    # Fire new_message_sent event and wait for reply of framer
+    async def handle_new_sent_message(self, data, context, eom):
+        self.loop.create_task(self.new_sent_message(data, context, eom))
+        data = await self.await_framer_send()
+        return data
+    # Fire handle_received_data event and wait for reply of framer
+    async def handle_received(self, connection):
+        self.loop.create_task(self.handle_received_data(connection))
+        context, data, eom = await self.await_framer_receive()
+        return context, data, eom
+
+    #  Declare a connection ready
+    def make_connection_ready(self, connection):
+        if self.start_waiter is not None:
+            self.start_waiter.set_result(None)
+        return
+
+    def fail_connection(self, error):
+        return
+
+    def prepend_protocol(self, framer):
+        return
+
+    def send(self, data):
+        # Send
+        if len(self.send_waiter_list) > 0:
+            self.send_waiter_list[0].set_result(data)
+            del self.send_waiter_list[0]
+        return
+
+    def parse(self, connection, min_incomplete_length, max_length):
+        return connection.recv_buffer.decode(), None, False
+
+    def advance_receive_cursor(self, connection, length):
+        connection.recv_buffer = connection.recv_buffer[length:]
+        return
+
+    def deliver_and_advance_receive_cursor(self, context, length, eom):
+        return
```

Comment:
I was unsure if this should be implemented as the combined call was missing the data argument in the draft and it was not clear to me wether or not that was a mistake or I didnt fully understand how it would get the data. It would not be an issue to implement if we assume it also has the data parameter.

Reply:
I suggest implementing it with the data parameter, since this is probably just a bug in the spec.